### ISSUE TITLE
List with optional value initialized with list multiplication is detected as error

### ISSUE THREAD

**Issue by ironfoot** [microsoft/pyright#1073]
**Describe the bug**
List with optional value when initialized with list multiplication is detected as error.

**To Reproduce**
```python
from typing import List, Optional

mylist: List[Optional[int]] = [None] * 10  # This displays error.
mylist: List[Optional[int]] = [1] * 10  # Same error.
mylist: List[Optional[int]] = [None, None, None]  # This is okay.
```

Error message
```
Expression of type "List[None]" cannot be assigned to declared type "List[int | None]"
  TypeVar "_T" is invariant
    Type "None" cannot be assigned to type "int | None"Pylance (reportGeneralTypeIssues)
```

**Expected behavior**
No error in all three cases.

**VS Code extension or command-line**
Pylance VS Code extension version: 2020.9.6

**Comment by erictraut** [comment]
Please refer to this issue: https://github.com/microsoft/pyright/issues/736

**Comment by erictraut** [comment]
As a general rule, I don't add special-case knowledge of internal behaviors of classes within the type checker — even for built-in classes like `list`. This might be one of those rare cases where an exception is justified. I'll think about it more.

**Comment by ironfoot** [comment]
I'm not entirely sure if I understood everything.
So, this is mostly for my learning but,

LHS type is determined from the annotation as `List[int | None]`.
RHS type is determined to be `List[None]` from the described steps.
Shouldn't it be a non-error to assign RHS into LHS since LHS type is a superset of RHS?

or maybe you meant `[None]` is determined as `List[None]` but `[None] * 2` cannot be determined as `List[None]` without the special-case knowledge?

**Comment by erictraut** [comment]
`List[None]` cannot be assigned to `List[int | None]`. Those are not compatible types. If you are interested in why, check out [this documentation](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#bidirectional-type-inference-expected-types).

Note that all of the following work without an error:
```python
from typing import List, Optional

mylist: Sequence[Optional[int]] = [None] * 10  # This displays error.
mylist: Sequence[Optional[int]] = [1] * 10  # Same error.
mylist: Sequence[Optional[int]] = [None, None, None]  # This is okay.
```

**Comment by ironfoot** [comment]
In terms of why it cannot be assigned, I'm still puzzled.
Bidirectional type inference describes how an ambiguity on RHS can be eliminated using the type of LHS.
In this case, isn't the type known for both sides and there is no ambiguity? That is, LHS is `List[int | None]` and RHS is `List[None]`. I didn't quite get how that leads to the two types not being incompatible.

I have one more question.
```python
mylist: List[Optional[int]] = [None, None, None]
```
In this case, RHS type is `List[None]` and it can be assigned to `List[int | None]`. Is this correct?
I didn't quite follow what is different between this and above when `List[None]` could not be assigned to `List[int | None]`.

**Comment by jakebailey** [comment]
Consider this code. If we allowed `y = x`, then `x` could have an `int` appended to it.

```python
x: List[None] = [None, None, None]
y: List[Optional[int]] = x

y.append(1234)
```

Sequence works because the definition allows for assignments in that direction (as the design of it takes into consideration that values are only read, covariance, etc).

In that `mylist: List[Optional[int]] = [None, None, None]` example, the entire right side is inferred, so the left side influences what the right is inferred to be.

**Comment by ironfoot** [comment]
I understand now.
Thanks for the explanation.

**Comment by erictraut** [comment]
I realized that I provided the wrong documentation link — or at least a link that wasn't as helpful as I had hoped. [Here](https://github.com/microsoft/pyright/blob/master/docs/type-concepts.md#generic-types) is what I had intended to link to.

**Comment by lazerkat** [comment]
Why would Pyright not infer the type of `[None]` as `Union[List[None], List[Optional[Any]]`, or even `Union[List[None], List[Optional[T]]` where `T` is dynamically inferred later? (I'm assuming these can be assigned to a `List[Optional[int]]`, which may not be true.)

The handling of Optional and None is strange in Python, because the type dynamically changes after an if guard from Optional[T] -> T. It seems like the typing of None has to be strange to accommodate this.

**Comment by erictraut** [comment]
Let me clarify a few things.

`Optional[X]` is just shorthand for `Union[None, X]`. It is treated like any other union.

Similarly, None is just another type from the perspective of the type checking logic. There's nothing strange or different about its handling.

You said "the type dynamically changes...". I think you're referring to what we call "conditional type narrowing". This is a technique whereby the type checker can determine that the value of an expression has a "narrower" type at certain places in the code flow graph because of conditional checks that affect the code flow. One technique for narrowing is removal of subtypes from within a union. Here again, None is treated like all other types. Consider the following code:

```python
def foo(a: Union[None, Tuple[()], int]):
    reveal_type(a) # Union[None, Tuple[()], int]
    if a:
        reveal_type(a) # int
```

You can see that `a` starts life as a union of three types, but within the `if` block, it has been narrowed to just type `int`. In this case, `None` and `Tuple[()]` (an empty tuple) are eliminated from the original union as part of the type narrowing process.

For more details about how pyright performs type inference for lists when there is no additional context (i.e. no possibility of bidirectional type inference), refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#list-expressions).

It would be a bad idea for any type inference operation to produce a union type because unions often cannot be assigned to other types. When a union is the source of an assignment, all of its subtypes must be assignable to the destination.

Finally, I'm not sure what you mean by "where `T` is dynamically inferred later". I think you are probably misunderstanding how type variables work. For a deeper understanding of type variables, refer to [PEP 483](https://www.python.org/dev/peps/pep-0483/) and [PEP 484](https://www.python.org/dev/peps/pep-0484/).

**Comment by lazerkat** [comment]
Thanks for the detailed reply, Eric.

My notation wasn't very clear; I was assuming that Pyright used something like Hindley-Milner type inference, and by `T` I meant an existential variable: a specific, unknown type. I was thinking the inference would look like:

```python
x: List[Optional[int]] = [None] * 5

# Type unification proceeds as:
None  -> Optional[T], some unknown type
[None] -> List[Optional[T]]
[None] * 5 -> returns a list of the same type, which must be List[Optional[T]], where T is the same T as before
List[Optional[int]] = List[Optional[T]] -> resolves T to int
```

With this approach, we would also deduce that the `None` literal was actually an `Optional[int]`. However, I had forgotten that `Optional` is just `Union[None, x]`, so this may be the problem (or there may be valid reasons not to treat `None` as an `Optional[T]`).

Stepping back, I feel like this expression _should_ type check properly, in a sane and well-defined type system. It's not clear to me if the issue is:

1. The Python's type system (which here must mean the type annotations, which probably still have trouble representing everything possible in Python)
2. Limitations in Pyright, like limited type inference for performance reasons or crazy Python corner cases

Finally, is there a good place to start understanding how Pyright's type deduction works, and what it's limitations are?

**Comment by erictraut** [comment]
Your description is correct for the first few steps of type unification. The problem occurs when you hit `[None] * 5`. A type checker can't assume "returns a list of the same type". The multiply operator (as with other binary operators) in Python cannot be assumed to return the same type as its operands. To analyze a multiply operator, a type checker must perform a complex set of steps including:

1. Fully resolve the types of the LHS and RHS operands
2. Look up the `__mul__` method on the LHS operand type (or possibly iterate over multiple types if the LHS is a union) and determine if `__mul__` is supported for the RHS type; if it is supported, use the specialized version of the return type
3. If `__mul__` is not supported for the RHS type, look up the `__rmul__` method on the RHS type and repeat the same process using the LHS type as an argument

A chicken-and-egg problem occurs because fully resolving the LHS and RHS types cannot be done without knowing `T`.

The best documentation I can offer you on Pyright's type deduction is [this](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md).

Here's [a link in the source](https://github.com/microsoft/pyright/blob/50d0e7b5074f2d7e4e427dac95c07509776688f8/packages/pyright-internal/src/analyzer/typeEvaluator.ts#L7484) to the code that performs the analysis for binary operations, if you're interested in digging in to that.
